File: pilot/src/generator/kv.rs
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};

use crate::storage::rocks::RocksStore;

use super::GenerateParams;
use super::types::OriginOp;

/// Key padding width — 20 digits for sorted numeric keys.
const KEY_PADDING: usize = 20;

/// Generates origin datasets for all KV-family profiles.
pub struct BasicKvGenerator {
    rng: StdRng,
    params: GenerateParams,
    /// Optional RocksStore for CAS version lookups.
    rocks: Option<RocksStore>,
    /// Tracks index keys for secondary index reads.
    index_keys: Vec<String>,
    /// Tracks sequence prefixes for sequence reads.
    sequence_prefixes: Vec<String>,
    op_counter: usize,
    /// Next key index to put during warmup phase.
    warmup_cursor: usize,
}

impl BasicKvGenerator {
    pub fn new(params: &GenerateParams) -> Self {
        Self {
            rng: StdRng::seed_from_u64(params.seed),
            params: params.clone(),
            rocks: None,
            index_keys: Vec::new(),
            sequence_prefixes: Vec::new(),
            op_counter: 0,
            warmup_cursor: 0,
        }
    }

    pub fn with_rocks(mut self, rocks: RocksStore) -> Self {
        self.rocks = Some(rocks);
        self
    }

    pub fn generate(mut self) -> Vec<OriginOp> {
        let mut ops = Vec::new();
        while self.op_counter < self.params.ops {
            ops.push(self.gen_op());
        }
        ops
    }

    /// Generate the next N ops. Can be called repeatedly for incremental generation.
    pub fn gen_batch(&mut self, count: usize) -> Vec<OriginOp> {
        let mut ops = Vec::new();
        for _ in 0..count {
            ops.push(self.gen_op());
        }
        ops
    }

    /// Generate a single op based on the full workload weights.
    /// During warmup, emits puts for all keys in the key space first.
    fn gen_op(&mut self) -> OriginOp {
        // Warmup: put every key in the key space before random ops
        if !self.params.skip_warmup && self.warmup_cursor < self.params.key_space.count {
            let id = self.next_id();
            let key = self.format_key(self.warmup_cursor);
            let value = self.random_value();
            self.warmup_cursor += 1;
            return OriginOp::put(id, key, value);
        }

        let id = self.next_id();
        let roll: f64 = self.rng.r#gen();
        let workload = self.params.resolved_workload();
        let total: f64 = workload.values().sum();
        if total <= 0.0 {
            return self.gen_put(id);
        }

        let mut cumulative = 0.0;
        for (op_type, weight) in &workload {
            cumulative += weight / total;
            if roll < cumulative {
                return match op_type.as_str() {
                    "put" => self.gen_put(id),
                    "get" => self.gen_get(id),
                    "get_floor" => self.gen_get_floor(id),
                    "get_ceiling" => self.gen_get_ceiling(id),
                    "get_lower" => self.gen_get_lower(id),
                    "get_higher" => self.gen_get_higher(id),
                    "delete" => self.gen_delete(id),
                    "delete_range" => self.gen_delete_range(id),
                    "list" => self.gen_list(id),
                    "range_scan" => self.gen_range_scan(id),

                    "ephemeral_put" => self.gen_ephemeral_put(id),
                    "indexed_put" => self.gen_indexed_put(id),
                    "indexed_get" => self.gen_indexed_get(id),
                    "indexed_list" => self.gen_indexed_list(id),
                    "indexed_range_scan" => self.gen_indexed_range_scan(id),
                    "sequence_put" => self.gen_sequence_put(id),
                    "watch_start" => self.gen_watch_start(id),
                    "session_restart" => self.gen_session_restart(id),
                    "get_notifications" => self.gen_get_notifications(id),
                    _ => self.gen_put(id),
                };
            }
        }
        self.gen_put(id)
    }

    // ── Write operations ──

    fn gen_put(&mut self, id: String) -> OriginOp {
        let key = self.random_key();
        let value = self.random_value();
        OriginOp::put(id, key, value)
    }

    fn gen_delete(&mut self, id: String) -> OriginOp {
        let key = self.random_key();
        OriginOp::delete(id, key)
    }

    fn gen_delete_range(&mut self, id: String) -> OriginOp {
        let ks = &self.params.key_space;
        let lo = self.rng.gen_range(0..ks.count);
        // Cap range to at most 5 keys
        let span = self.rng.gen_range(1..=5.min(ks.count));
        let hi = (lo + span).min(ks.count);
        OriginOp::delete_range(id, self.format_key(lo), self.format_key(hi))
    }

    fn gen_ephemeral_put(&mut self, id: String) -> OriginOp {
        let key = self.random_key();
        let value = self.random_value();
        OriginOp::ephemeral_put(id, key, value)
    }

    fn gen_indexed_put(&mut self, id: String) -> OriginOp {
        let key = self.random_key();
        let value = self.random_value();
        let idx = &self.params.index;
        let index_key_num = self.rng.gen_range(0..idx.key_count);
        let index_key = format!("idx-{index_key_num:04}");

        if !self.index_keys.contains(&index_key) {
            self.index_keys.push(index_key.clone());
        }

        OriginOp::indexed_put(id, key, value, idx.name.clone(), index_key)
    }

    fn gen_sequence_put(&mut self, id: String) -> OriginOp {
        let ks = &self.params.key_space;
        let prefix = format!("{}seq/", ks.prefix);
        let value = self.random_value();

        if !self.sequence_prefixes.contains(&prefix) {
            self.sequence_prefixes.push(prefix.clone());
        }

        OriginOp::sequence_put(id, prefix, value, 1)
    }

    // ── Read operations ──

    fn gen_get(&mut self, id: String) -> OriginOp {
        OriginOp::get(id, self.random_key())
    }

    fn gen_get_floor(&mut self, id: String) -> OriginOp {
        OriginOp::get_floor(id, self.random_key())
    }

    fn gen_get_ceiling(&mut self, id: String) -> OriginOp {
        OriginOp::get_ceiling(id, self.random_key())
    }

    fn gen_get_lower(&mut self, id: String) -> OriginOp {
        OriginOp::get_lower(id, self.random_key())
    }

    fn gen_get_higher(&mut self, id: String) -> OriginOp {
        OriginOp::get_higher(id, self.random_key())
    }

    fn gen_range_scan(&mut self, id: String) -> OriginOp {
        let ks = &self.params.key_space;
        let a = self.rng.gen_range(0..ks.count);
        let b = self.rng.gen_range(0..ks.count);
        let lo = a.min(b);
        let hi = a.max(b) + 1;
        OriginOp::range_scan(id, self.format_key(lo), self.format_key(hi))
    }

    fn gen_list(&mut self, id: String) -> OriginOp {
        let ks = &self.params.key_space;
        // List all keys in the run prefix
        OriginOp::list(id, self.format_key(0), format!("{}~", ks.prefix))
    }

    fn gen_indexed_get(&mut self, id: String) -> OriginOp {
        let idx = &self.params.index;
        if self.index_keys.is_empty() {
            return self.gen_get(id);
        }
        let i = self.rng.gen_range(0..self.index_keys.len());
        let index_key = self.index_keys[i].clone();
        OriginOp::indexed_get(id, idx.name.clone(), index_key)
    }

    fn gen_indexed_list(&mut self, id: String) -> OriginOp {
        let idx = &self.params.index;
        let a = self.rng.gen_range(0..idx.key_count);
        let b = self.rng.gen_range(0..idx.key_count);
        let lo = a.min(b);
        let hi = a.max(b) + 1;
        OriginOp::indexed_list(id, idx.name.clone(), format!("idx-{lo:04}"), format!("idx-{hi:04}"))
    }

    fn gen_indexed_range_scan(&mut self, id: String) -> OriginOp {
        let idx = &self.params.index;
        let a = self.rng.gen_range(0..idx.key_count);
        let b = self.rng.gen_range(0..idx.key_count);
        let lo = a.min(b);
        let hi = a.max(b) + 1;
        OriginOp::indexed_range_scan(id, idx.name.clone(), format!("idx-{lo:04}"), format!("idx-{hi:04}"))
    }

    // ── Session & notification ops ──

    fn gen_watch_start(&mut self, id: String) -> OriginOp {
        OriginOp::watch_start(id, self.params.key_space.prefix.clone())
    }

    fn gen_session_restart(&mut self, id: String) -> OriginOp {
        OriginOp::session_restart(id)
    }

    fn gen_get_notifications(&mut self, id: String) -> OriginOp {
        OriginOp::get_notifications(id)
    }

    // ── Helpers ──

    fn next_id(&mut self) -> String {
        self.op_counter += 1;
        format!("op-{:04}", self.op_counter)
    }

    /// Format a key with 20-digit padding under the run prefix.
    fn format_key(&self, n: usize) -> String {
        format!("{}{:0width$}", self.params.key_space.prefix, n, width = KEY_PADDING)
    }

    fn random_key(&mut self) -> String {
        let n = self.rng.gen_range(0..self.params.key_space.count);
        self.format_key(n)
    }

    fn random_value(&mut self) -> String {
        let vc = &self.params.value;
        let charset = b"abcdefghijklmnopqrstuvwxyz0123456789";
        let len = self.rng.gen_range(vc.min_len..=vc.max_len);
        let s: String = (0..len)
            .map(|_| {
                let idx = self.rng.gen_range(0..charset.len());
                charset[idx] as char
            })
            .collect();
        format!("{}{s}", vc.prefix)
    }
}

File: pilot/generator/kv.go
package generator

import (
	"fmt"
	"math/rand"
)

// keyPadding is the width for zero-padded numeric keys.
const keyPadding = 20

const (
	secondaryIndexPutRatio       = 0.25 / 0.40
	secondaryIndexGetRatio       = 0.15 / 0.25
	secondaryIndexListRatio      = 0.10 / 0.15
	secondaryIndexRangeScanRatio = 0.10 / 0.15
	sequencePutRatio             = 0.40 / 0.50
)

// BasicKvGenerator generates origin datasets for all KV-family profiles.
type BasicKvGenerator struct {
	rng              *rand.Rand
	params           GenerateParams
	indexKeys        []string
	sequencePrefixes []string
	opCounter        int
	warmupCursor     int
}

// NewBasicKvGenerator creates a new BasicKvGenerator from the given params.
func NewBasicKvGenerator(params *GenerateParams) *BasicKvGenerator {
	p := *params
	// Guard against zero values that would cause Intn(0) panics
	if p.KeySpace.Count <= 0 {
		p.KeySpace.Count = 20_000
	}
	if p.Index.KeyCount <= 0 {
		p.Index.KeyCount = 50
	}
	if p.Value.MinLen <= 0 {
		p.Value.MinLen = 4
	}
	if p.Value.MaxLen <= 0 {
		p.Value.MaxLen = 12
	}
	return &BasicKvGenerator{
		rng:    rand.New(rand.NewSource(int64(p.Seed))),
		params: p,
	}
}

// GenBatch generates the next count operations.
func (g *BasicKvGenerator) GenBatch(count int) []OriginOp {
	ops := make([]OriginOp, 0, count)
	for i := 0; i < count; i++ {
		ops = append(ops, g.genOp())
	}
	return ops
}

// Generate generates all operations up to params.Ops and returns them.
func (g *BasicKvGenerator) Generate() []OriginOp {
	var ops []OriginOp
	for g.opCounter < g.params.Ops {
		ops = append(ops, g.genOp())
	}
	return ops
}

// genOp generates a single op based on the full workload weights.
// During warmup, emits puts for all keys in the key space first.
func (g *BasicKvGenerator) genOp() OriginOp {
	// Warmup: put every key in the key space before random ops
	if !g.params.SkipWarmup && g.warmupCursor < g.params.KeySpace.Count {
		id := g.nextID()
		key := g.formatKey(g.warmupCursor)
		value := g.randomValue()
		g.warmupCursor++
		return Put(id, key, value)
	}

	id := g.nextID()
	roll := g.rng.Float64()
	workload := g.params.ResolvedWorkload()

	total := 0.0
	for _, w := range workload {
		total += w
	}
	if total <= 0.0 {
		return g.genPut(id)
	}

	cumulative := 0.0
	for opType, weight := range workload {
		cumulative += weight / total
		if roll < cumulative {
			return g.dispatchOp(opType, id)
		}
	}
	return g.genPut(id)
}

func (g *BasicKvGenerator) dispatchOp(opType, id string) OriginOp {
	switch opType {
	case "put":
		return g.genProfilePut(id)
	case "get":
		return g.genProfileGet(id)
	case "get_floor":
		return g.genGetWithComparison(id, "floor")
	case "get_ceiling":
		return g.genGetWithComparison(id, "ceiling")
	case "get_lower":
		return g.genGetWithComparison(id, "lower")
	case "get_higher":
		return g.genGetWithComparison(id, "higher")
	case "delete":
		return g.genDelete(id)
	case "delete_range":
		return g.genDeleteRange(id)
	case "list":
		return g.genProfileList(id)
	case "range_scan":
		return g.genProfileRangeScan(id)
	case "watch_start":
		return g.genWatchStart(id)
	case "session_restart":
		return g.genSessionRestart(id)
	case "get_notifications":
		return g.genGetNotifications(id)
	default:
		return g.genPut(id)
	}
}

func (g *BasicKvGenerator) genProfilePut(id string) OriginOp {
	switch g.params.Generator {
	case "kv-ephemeral-notification":
		return g.genEphemeralPut(id)
	case "kv-secondary-index":
		if g.rng.Float64() < secondaryIndexPutRatio {
			return g.genIndexedPut(id)
		}
	case "kv-sequence":
		if g.rng.Float64() < sequencePutRatio {
			return g.genSequencePut(id)
		}
	}
	return g.genPut(id)
}

func (g *BasicKvGenerator) genProfileGet(id string) OriginOp {
	if g.params.Generator == "kv-secondary-index" && g.rng.Float64() < secondaryIndexGetRatio {
		return g.genIndexedGet(id)
	}
	return g.genGet(id)
}

func (g *BasicKvGenerator) genProfileList(id string) OriginOp {
	if g.params.Generator == "kv-secondary-index" && g.rng.Float64() < secondaryIndexListRatio {
		return g.genIndexedList(id)
	}
	return g.genList(id)
}

func (g *BasicKvGenerator) genProfileRangeScan(id string) OriginOp {
	if g.params.Generator == "kv-secondary-index" && g.rng.Float64() < secondaryIndexRangeScanRatio {
		return g.genIndexedRangeScan(id)
	}
	return g.genRangeScan(id)
}

// -- Write operations --

func (g *BasicKvGenerator) genPut(id string) OriginOp {
	return Put(id, g.randomKey(), g.randomValue())
}

func (g *BasicKvGenerator) genDelete(id string) OriginOp {
	return Delete(id, g.randomKey())
}

func (g *BasicKvGenerator) genDeleteRange(id string) OriginOp {
	ks := &g.params.KeySpace
	lo := g.rng.Intn(ks.Count)
	maxSpan := 5
	if ks.Count < maxSpan {
		maxSpan = ks.Count
	}
	span := g.rng.Intn(maxSpan) + 1
	hi := lo + span
	if hi > ks.Count {
		hi = ks.Count
	}
	return DeleteRange(id, g.formatKey(lo), g.formatKey(hi))
}

func (g *BasicKvGenerator) genEphemeralPut(id string) OriginOp {
	return EphemeralPut(id, g.randomKey(), g.randomValue())
}

func (g *BasicKvGenerator) genIndexedPut(id string) OriginOp {
	key := g.randomKey()
	value := g.randomValue()
	idx := &g.params.Index
	indexKeyNum := g.rng.Intn(idx.KeyCount)
	indexKey := fmt.Sprintf("idx-%04d", indexKeyNum)

	found := false
	for _, ik := range g.indexKeys {
		if ik == indexKey {
			found = true
			break
		}
	}
	if !found {
		g.indexKeys = append(g.indexKeys, indexKey)
	}

	return IndexedPut(id, key, value, idx.Name, indexKey)
}

func (g *BasicKvGenerator) genSequencePut(id string) OriginOp {
	prefix := fmt.Sprintf("%sseq/", g.params.KeySpace.Prefix)
	value := g.randomValue()

	found := false
	for _, sp := range g.sequencePrefixes {
		if sp == prefix {
			found = true
			break
		}
	}
	if !found {
		g.sequencePrefixes = append(g.sequencePrefixes, prefix)
	}

	return SequencePut(id, prefix, value, int64(1))
}

// -- Read operations --

func (g *BasicKvGenerator) genGet(id string) OriginOp {
	return Get(id, g.randomKey())
}

func (g *BasicKvGenerator) genGetWithComparison(id, comparison string) OriginOp {
	return GetWithComparison(id, g.randomKey(), comparison)
}

func (g *BasicKvGenerator) genRangeScan(id string) OriginOp {
	const maxRange = 100
	ks := &g.params.KeySpace
	lo := g.rng.Intn(ks.Count)
	span := g.rng.Intn(min(maxRange, ks.Count)) + 1
	hi := lo + span
	if hi > ks.Count {
		hi = ks.Count
	}
	return Scan(id, g.formatKey(lo), g.formatKey(hi))
}

func (g *BasicKvGenerator) genList(id string) OriginOp {
	const maxRange = 100
	ks := &g.params.KeySpace
	lo := g.rng.Intn(ks.Count)
	span := g.rng.Intn(min(maxRange, ks.Count)) + 1
	hi := lo + span
	if hi > ks.Count {
		hi = ks.Count
	}
	return List(id, g.formatKey(lo), g.formatKey(hi))
}

func (g *BasicKvGenerator) genIndexedGet(id string) OriginOp {
	if len(g.indexKeys) == 0 {
		return g.genGet(id)
	}
	i := g.rng.Intn(len(g.indexKeys))
	indexKey := g.indexKeys[i]
	return IndexedGet(id, g.params.Index.Name, indexKey)
}

func (g *BasicKvGenerator) genIndexedList(id string) OriginOp {
	const maxRange = 100
	idx := &g.params.Index
	lo := g.rng.Intn(idx.KeyCount)
	span := g.rng.Intn(min(maxRange, idx.KeyCount)) + 1
	hi := lo + span
	if hi > idx.KeyCount {
		hi = idx.KeyCount
	}
	return IndexedList(id, idx.Name, fmt.Sprintf("idx-%04d", lo), fmt.Sprintf("idx-%04d", hi))
}

func (g *BasicKvGenerator) genIndexedRangeScan(id string) OriginOp {
	const maxRange = 100
	idx := &g.params.Index
	lo := g.rng.Intn(idx.KeyCount)
	span := g.rng.Intn(min(maxRange, idx.KeyCount)) + 1
	hi := lo + span
	if hi > idx.KeyCount {
		hi = idx.KeyCount
	}
	return IndexedRangeScan(id, idx.Name, fmt.Sprintf("idx-%04d", lo), fmt.Sprintf("idx-%04d", hi))
}

// -- Session & notification ops --

func (g *BasicKvGenerator) genWatchStart(id string) OriginOp {
	return WatchStart(id, g.params.KeySpace.Prefix)
}

func (g *BasicKvGenerator) genSessionRestart(id string) OriginOp {
	return SessionRestart(id)
}

func (g *BasicKvGenerator) genGetNotifications(id string) OriginOp {
	return GetNotifications(id)
}

// -- Helpers --

func (g *BasicKvGenerator) nextID() string {
	g.opCounter++
	return fmt.Sprintf("op-%04d", g.opCounter)
}

func (g *BasicKvGenerator) formatKey(n int) string {
	return fmt.Sprintf("%s%0*d", g.params.KeySpace.Prefix, keyPadding, n)
}

func (g *BasicKvGenerator) randomKey() string {
	n := g.rng.Intn(g.params.KeySpace.Count)
	return g.formatKey(n)
}

func (g *BasicKvGenerator) randomValue() string {
	vc := &g.params.Value
	charset := "abcdefghijklmnopqrstuvwxyz0123456789"
	length := vc.MinLen + g.rng.Intn(vc.MaxLen-vc.MinLen+1)
	b := make([]byte, length)
	for i := range b {
		b[i] = charset[g.rng.Intn(len(charset))]
	}
	return vc.Prefix + string(b)
}
